### ISSUE TITLE
Maximizing good results and Minimizing irrelevant results

### ISSUE THREAD

**Issue by dianos** [typesense/typesense#312]
We are trying to get good settings for searching thousands of items which would generate minimal irrelevant results.
Example searches with different settings for drop tokens, num typos, and typo tokens. 

`   queryBy: "description,name,categories,attrib",
    num_typos: gtypo,
    drop_tokens_threshold: gdrop,
    typo_tokens_threshold: gttoken,
`

**Example 1**:  Searching for this Item with Drop tokens set to 0:
`IFP5550-BUN` Viewsonic Viewboard Ifp5550 Collaboration Display - 55 Lcd - Arm Cortex A53 1.20 Ghz - 2 Gb - Infrared (irda) - Touchscreen - 16:9 Aspect Ratio - 3840 X 2160 - Led - 350 Nit - 1,200:1 Contrast Ratio - 2160p - Usb - Hdmi - Vga - Android 5.1 Lollipop Ez Bu`

a) When searching for "IFP5550BUN" the item is found, but when searching for  "IFP5550BUN Viewsonic" it is no longer found.
b) When searching for "ifp5550 viewsonic bun" with typo tokens disabled it finds the above 1 item but when trying to filter that results on a Brand for "viewsonic", more items are added to the search results.

-----------  
**Example 2**:
Having these 3 Items:  "SPZ007 space", "SPZ008 spacepole", "SPZ005 space"
a) when searching for "SPZ00":  items   SPZ008, SPZ005 are found but item SPZ007 is not found when typo tokens are set to 0, but found with typo tokens set to more then 0
b) when searching for "spz00 space" no items are found at all with settings set to 0.
c) when searching for "spz00 space",with num typos set to 0, drop tokens set to more than 0 and typos kept at 1: thousands of 'irrelevant' items are found that have the word "space" in them but no mention of "spz00"
d)when num typos is increased from 0 to 1 there is less results found in the above search not more. 
e) when searching for "spz00 space" having typos set to 1, and droptokens to 1 no items are found.  But when typotokens are increased as well to more then one: hundreds of items are found such as "Sp-800, Spare" that seem quite distant from  "spz00 space" 
f) when num typos are increased further to 3, for the above search, there is hundreds LESS results but they seem even more irrelevant with matches such us "TSP700II" highlighted 10 items down the line but first items don't even have a highlight on a matching word.

**Comment by dianos** [comment]
From my understanding num_typos should control the amount of misspellings in a given word.  Yet having an item named "RV345P-K9-NA"  when searching for "V345k9N" nothing is found unless ttypotokens are set to more then 1.  

Also when searching for "RV345P-K9-NA" nothing is found unless num_typos is set to 2 or more, why would the minus inside the search string be treated as a missspelling?  shouldn't it be space minus " -" to be treated as a NOT the following word to begin with?

**Comment by kishorenc** [comment]
@dianos 

> When searching for "IFP5550BUN" the item is found, but when searching for "IFP5550BUN Viewsonic" it is no longer found.

In the demo link you had shared, both those queries are giving me the expected item at position 1. 

> When searching for "ifp5550 viewsonic bun" with typo tokens disabled it finds the above 1 item but when trying to filter that results on a Brand for "viewsonic", more items are added to the search results.

This is because during indexing, "IFP5550-BUN" is indexed as a single word "IFP5550BUN" and so when you have `bun` as a separate keyword in search, it tries to find other documents also if you have enabled `drop_tokens_threshold`. 

> a) when searching for "SPZ00": items SPZ008, SPZ005 are found but item SPZ007 is not found when typo tokens are set to 0, but found with typo tokens set to more then 0

I think this will be fixed as part of a recent build here: https://github.com/typesense/typesense/issues/309#issuecomment-876356595 -- can you please try against that locally? I will be happy to go through the other examples once you've had a chance to try out this new build.

> Yet having an item named "RV345P-K9-NA" when searching for "V345k9N" nothing is found

This is because `RV345P-K9-NA` is indexed as `RV345PK9NA`. So `V345k9N` will not be matched to that since it's not within 2 typos from that token.

> Also when searching for "RV345P-K9-NA" nothing is found unless num_typos is set to 2 or more,

This is also addressed in the RC build. May I know what version of Typesense you are using?

**Comment by dianos** [comment]
![viewsonic2](https://user-images.githubusercontent.com/4541410/125428900-eec0fa24-c939-418c-a2fb-0e4acba6fea6.jpg)

**Comment by kishorenc** [comment]
The latest Docker RC build is: `typesense/typesense:0.21.0.rc24`. You don't have to re-index the data, just restarting Typesense on the RC version against existing data directory is all you need to do. I will appreciate if you can give that a spin and see if you notice any improvements. We will be releasing this RC as final 0.21 release in a day or two.

**Comment by dianos** [comment]
unfortunately ?query=Viewsonic%20IFP5550BUN still and issue with the latest build (0.21.0.rc24) of typesense.  Both drop tokens and typo tokens have to be more then 1 to get any results at all, even though these are exact keywords.  
If I increase num typos to 1 from 0, I also get less results by about 40 for this case.
![viewsonic3](https://user-images.githubusercontent.com/4541410/125452457-600aa54e-5417-4700-a0f2-ad1cf57f1574.jpg)

**Comment by dianos** [comment]
@kishorenc  but the real problem in this case from usability perspective is that drop tokens with the minimal setting actually drops the whole keyword such as "IFP5550BUN" and returns hundreds of irrelevant results.  while with drop tokens 0 it returns nothing.

**Comment by kishorenc** [comment]
@dianos Can you please provide me a minimally reproduceable dataset that exhibits the problem? For example, I tried indexing a single document and querying as you have described, but it is providing the result to me:

https://gist.github.com/kishorenc/8cc793a873e11f687647c9aac0193568

If you can share a minimally reproduceable sample, then it will be easier for us to figure out what is going wrong. Certainly want to get to the bottom of this issue.

**Comment by dianos** [comment]
@kishorenc yes will give you a minimal json file.  but one thing i wanted to note: this particular issue seems to occur when searching across different fields.  so i have IFP5550BUN in 'name' and Viewsonic in 'description' fields.   if IFP5550BUN is also present in the description field then it comes up with everything set to 0.

**Comment by dianos** [comment]
javascript import , and json with 23 viewsonic items.  I tried changing the order of fields searched description vs name but it makes no difference.  Let me know if you need anything else to reproduce.
[viewsonic.zip](https://github.com/typesense/typesense/files/6810899/viewsonic.zip)

**Comment by dianos** [comment]
json for the issue where spz007 is not found when searching for spz00; but spz005 and spz008 are found.
[spz.zip](https://github.com/typesense/typesense/files/6811033/spz.zip)
------------
However I think the larger issue is why there is such a big difference in results between settings of 0 and 1 for drop tokens; maybe there needs to be a better scale where for instance only one character is dropped from the search query for the minimal setting not an entire word?

**Comment by dianos** [comment]
here is another strange one:
item description:
`Cisco RV345 Router - 18 Ports - Management Port - SlotsGigabit Ethernet - Rack-mountable`
when searching for `RV345 cisco 18` is not found, but when you search for `RV345 cisco 18 ports` it is found.   irrelevant of the search sensitivity settings.

**Comment by kishorenc** [comment]
@dianos 

Let's first look at the viewsonic example. I downloaded your sample data with indexing script and ran it. However, when I query for `IFP5550 Viewsonic`, I'm finding that record. Here's the query that I tried:

```
?q=IFP5550%20Viewsonic&query_by=description,name,categories,attrib&drop_tokens_threshold=0
```

Can you please provide me with a query on this dataset of 23 viewsonic documents which is unexpected?

**Comment by dianos** [comment]
Sorry good catch, i filtered on the wrong field...    The example you tried works but "IFP5550E1" is also in the description field if u were to remove it i bet it would stop working. From the data you have you can try "VBPEN002" that will come up but if u add Viewsonic it will no longer come up.   Alternatively here is json for IFP5550-BUN.
` {
        "name": "IFP5550-BUN",
        "image": [
            "https://www.directdial.com/dd2/img/item/A-150x150/704600.jpg"
        ],
        "description": "Viewsonic Viewboard Ifp5550 Collaboration Display - 55 Lcd - Arm Cortex A53 1.20 Ghz - 2 Gb - Infrared (irda) - Touchscreen - 16:9 Aspect Ratio - 3840 X 2160 - Led - 350 Nit - 1,200:1 Contrast Ratio - 2160p - Usb - Hdmi - Vga - Android 5.1 Lollipop Ez Bu",
        "description_fr": "Viewsonic Viewboard Ifp5550 Collaboration Display - 55 Lcd - Arm Cortex A53 1.20 Ghz - 2 Gb - Infrared (irda) - Touchscreen - 16:9 Aspect Ratio - 3840 X 2160 - Led - 350 Nit - 1,200:1 Contrast Ratio - 2160p - Usb - Hdmi - Vga - Android 5.1 Lollipop Ez Bu",
        "brand": "Viewsonic",
        "price": 6973,
        "categories": [
            "Computer Displays",
            "Electronic Writing Boards",
            "Interactive Whiteboard",
            "Viewsonic"
        ],
        "url_breadcrumb": "\/computer-displays\/electronic-writing-boards\/interactive-whiteboard\/viewsonic\/",
        "breadcrumb": "Computer Displays|Electronic Writing Boards|Interactive Whiteboard|Viewsonic",
        "popularity": 2,
        "instock": "false",
        "attrib": [
            "Product Type > Interactive Whiteboard"
        ]
    }`

**Comment by kishorenc** [comment]
>  From the data you have you can try "VBPEN002" that will come up but if u add "Viewsonic" it will no longer come up. 

@dianos Thanks, this is reproducible and I explain what is happening here:

Typesense searches for a query against individual fields (since the index is maintained at a per-field level) and then aggregates results across fields by combining the per-field matches and scores. In this specific example, `VBPEN002` is present only in the name, while `Viewsonic` is present only in the description. 

```
"name": "VB-PEN-002"
"description": "Viewsonic IFP, ViewBoard Passive Touch Pen x 2 (Double Tips), Iron, Black - Iron - Black - Interactive Display Device Supported",
```

When you search for `VB-PEN-002 viewsonic`, the search is made individually first against the `name` field and then against the `description` field. In both cases, there will be no results because `drop_tokens_threshold` is disabled since it is set to `0`. 

Question for you: what problem do you face when setting `drop_tokens_threshold: 1`? I realize that this can pull up generic `viewsonic` related results, but these will be ranked only **after** the record containing both `viewsonic` and `VB-PEN-002 ` . 

An alternative is to have an aggregated field that concatenates the text of all text fields you want to search on as a single field. Then you will `query_by` on this single concatenated field, but highlight the primary `name` and `description` fields of the document by sending `?highlight_fields=name,title` in the search request (this custom highlight feature is available only in the RC build). Would that help?

**Comment by dianos** [comment]
The problem with drop token setting is that it seems to be way too insensitive, setting between 0-1 can vary results drastically.   If you look at the example "RV345 cisco 18".  With drop tokens set to 1, that item comes up as 50 in the ranking.

**Comment by dianos** [comment]
@kishorenc I agree with you that having drop tokens enabled is desirable in the end.  But scoring of relevant items would need to be much better.  Let me know if you need json for the "cisco" example above, as this illustrates the problem with drop tokens very well. You have items with just one short word such as "18" scoring over items with 3 matching words.  While those items shouldn't even be coming up in the results at all.

**Comment by kishorenc** [comment]
Yes can you please provide sample dataset for this 'cisco' example?

**Comment by dianos** [comment]
So here we have two techs cisco and nexan.  Searching for "RV345 cisco 18" over "description, name".   You will see nexan items with just "18" matching over the much more relevant cisco items.
[e.zip](https://github.com/typesense/typesense/files/6814155/e.zip)

**Comment by dianos** [comment]
@kishorenc actually sorry, I'm searching over "description, name, categories, attrib"

**Comment by kishorenc** [comment]
I just saw your edited example involving `VBPEN002 accessories`. 

> Maybe because VB-PEN-003 is also in the description? 

Correct, the reason why `VB-PEN-003` shows up before `VB-PEN-002` is that Typesense scores matches on 3 fields higher than 2-field match. 

1. Both `VB-PEN-003` and `VB-PEN-002` records contain the `accessories` in the `categories` field
2. The `VB-PEN-003` record has the token `VB-PEN-003` in both name and description fields, while `VB-PEN-002` record contains the token `VB-PEN-002` ONLY in the `name` field. 

I think this issue is specific to model or part numbers, since a single character change can mean a totally different product.

I think the following approach will help:

1. In both `name` and `description` identify words that contain both characters and numbers (i.e. alphanumerical words), and move them into a separate `model` field. This need not be exact and can contain false positives.
2. Now when you search try using the model field first in `query_by` AND also disable typo tolerance on model field like this: `num_typos=0,1,1` if your `query_by` is `model,name,description`.

This way, model number is matched exactly and also treated as a higher priority match since it is further ahead in the field list. I'm not 100% sure that this work, but something along these lines will help here. 

From Typesense perspective, maybe we can add an option to automatically disable typos on any alphanumerical token. This will make this much easier.

**Comment by dianos** [comment]
I will redo testing with your model suggestion.  Can we look at the spz00 issue?  I provided 3 item json  for that.  Why would we need typo tokens to have an exact match appear in the results?

**Comment by kishorenc** [comment]
I think the `spz00` issue might be because of https://github.com/typesense/typesense/issues/302#issuecomment-872358157 -- can you take a look at that to see if it might apply here? We will have a fix for this soon. 

> Searching for "RV345 cisco 18" over "description, name". You will see nexan items with just "18" matching over the much more relevant cisco items.

Unable to reproduce that. 

See the script and response here: https://gist.github.com/kishorenc/ec82107f495f034aa2ba0d9ddb0fc8fa

The JSON dataset you linked to for this example had some issues, so I extracted out the fields to JSONL format so I could import above. The JSONL is here: https://gist.githubusercontent.com/kishorenc/3502a7289134baa9dfc76e8a5ed5520d/raw/e126567b4bd6e8e111411fbf2fcc62ae18da78d1/cisco.jsonl

**Comment by dianos** [comment]
`spz00` is 5 characters that is not a short prefix? I do have a `popularity` sort field defined.  What is very strange is that 2 of 3 items are found even with numtypos set to 0 (yet in other cases a nymTypos 1 setting will still not find an alphanumeric item# that has 1 character difference).  
Now we can't disable neither `drop tokens` nor `typo tokens`, as that will filter out exact and/or very close substring matches.

--------------
Regarding "RV345 cisco 18" over "description, name, categories, attrib".  it's actually `attrib` that is causing the largest issue.  What it looks like is that items that have "18" mentioned several times are scored higher then those that have all the keywords matched.  For instance "Cisco RV345P Router - 18" comes up 6th behind 4 irrelevant items in the dataset provided. 

After I re-indexed: `E182-72/4` is no longer coming up as a first match. But several items with only a matching "18" still come up ahead of much better matches.   The `attrib` field for `E182-72/4` has no mention of neither `RV345` nor `cisco`

E182-72/4 attrib value:
`["Product Type > SAN Storage System","Form Factor > Rack-mountable","Controller Type > Serial ATA: :Serial Attached SCSI (SAS)","Fibre Channel > Yes","RAID Supported > Yes","Maximum Power Supply Wattage > 1.20 kW: :1200 W","Product Family > E","Total Hard Drive Capacity Installed > 72 TB","Number of Hard Drives Supported > 18","Number of Hard Drives Installed > 18","Number of Solid State Drive Supported > 18","Drive Interface > Serial ATA"]`

--------------
Regarding the JSON import 
The culprit is the if statement on  line 147  in apicall.js.   What happens is that node.js is unable to allocate memory inside the if statement over what is currently the max allocated for node.js by the OS.   It needs to be changed to something that doesn't cause a large memory allocation inside an if check something more like:
if ( (typeof bodyParameters !== "undefined") && (Object.keys(bodyParameters).length !== 0) ) {
adding this statement was discussed here:
https://github.com/typesense/typesense-js/issues/38
once that statement is removed you can safely import hundreds of thousands of items in json format.

**Comment by dianos** [comment]
@kishorenc regarding the cisco+nexsan  dataset my multiples of "18" theory might not be correct as searching for "Cisco 18" actually brings in relevant "Cisco 18"  items on top; but adding other keywords brings in the irrelevant.  Searching for `Cisco 18 RV345P` brings in `nexsan` items, or searching for `Cisco 18 SlotsGigabit` will also have `SlotsGigabit` items interjected with irrelevant `nexsan` items.   why is this happening?

**Comment by kishorenc** [comment]
For the Cisco query, can you please give the exact query along with whether you are using num typos etc.? I will try reindexing again and querying.

**Comment by dianos** [comment]
`Cisco SlotsGigabit 18` this is a very concerning one.  I changed from `Cisco 18 SlotsGigabit`  which at least brought a couple cisco items on top.  but if you put 18 at the end it brings totally irrelevant results.  this is on port 3001 with 350K items.
<img width="960" alt="Cisco" src="https://user-images.githubusercontent.com/4541410/125624413-d23a6c44-3a32-469a-b01c-d5366fd17942.png">

**Comment by dianos** [comment]
>>For the Cisco query, can you please give the exact query along with whether you are using num typos etc.? I will try reindexing again and querying.
Well the problem as mentioned at the beginning is that setting drop tokens to 0 and/or typotokens to 0 will omit exact or substring matches.  While setting drop tokens to 1 brings totally irrelevant items in the top matches.   And as you can see adding a 3rd search keyword is really messing up the results.

**Comment by dianos** [comment]
Sorry that screenshot example was my bad, it was sorted by popularity not the score. --deleted
To reproduce some of the irrelevant results you can search for `RV345 cisco 18` with typo/drop settings to 1. You will see irrelevant items in the top 10 results.  While `Cisco RV345P Router - 18` which is a close match comes up down the list.

**Comment by dianos** [comment]
@kishorenc Actually that popularity sorted screenshot illustrates how broken `drop tokens` are from another  angle.  Here is the same query with the 2nd and 3rd parameter reversed. One brings in 57 results the other 5.7K. If we want to allow sorting by price for instance searching for 'Cisco SlotsGigabit 18 ' brings in totally irrelevant items thanks to 'drop tokens'.

<img width="570" alt="18_18_2nd" src="https://user-images.githubusercontent.com/4541410/125630419-323c073d-83b8-44e8-af4a-8a670d980042.png">
<img width="521" alt="18_18" src="https://user-images.githubusercontent.com/4541410/125630449-2f99fa59-db14-41ce-ab71-78021eb9f859.png">

**Comment by kishorenc** [comment]
@dianos 

It's getting a bit difficult to debug on Github -- are you on our Slack channel? I think it might be easier to chat over there. We can pick one particular case and get to the bottom of it.

**Comment by dianos** [comment]
Most of these issues are now fixed in  **typesense:0.22 RC** builds. 

Making one field that combines data from all fields that we search on solves issues with drop tokens, since these can then be set to 0.

Still keeping this open, for testing of upcoming changes.  And will add related issue regarding popularity based ordering.

**Comment by jasonbosco** [comment]
Released in v0.22. Could you open a new issue for issues found from here on?
